### ISSUE TITLE
low energy neutrons

### ISSUE THREAD

**Issue by cipriangal** [JeffersonLab/remoll#248]
### Environment: (where does this bug occur, have you tried other environments)
- branch (master for latest released): feature-add-SBSbunker
- revision (HEAD for most recent): commit 9e9c31101f122a349474f7bac735f270309880f2
- OS or system: MacOS, Linux
- Special ROOT or Geant4 versions? G4.10.04.p1

### Steps to reproduce: (give a step by step account of how to trigger the bug)
1. run macros/runHallDet.mac with "/remoll/physlist/register QGSP_BERT_HP" (I ran about 1e7 events). Look at detector 101 at the roof (or detector 5555 right above the target)

Plotting the neutron spectrum shows odd behaviour:
### Expected Result: (what do you expect when you execute the steps above)
The right plot is from a prex simulation with 1 GeV beam looking at a similar detector on the roof (with the same physics list, ran with earlier version of G4.10.2.2). We can see similar features up to about 0.01 MeV, however the thermal neutrons below seem to be completely missing from the remoll simulation. 

### Actual Result: (what do you get when you execute the steps above)
The left plot.

![image](https://user-images.githubusercontent.com/8987591/55243351-6ee80100-5215-11e9-8251-bda445c9f300.png)

**Comment by wdconinc** [comment]
Two questions:
- hit.e is the total energy, including restmass, so for massive particles it has a non-zero lower limit even when the particle is at rest. I don't think this is an issue here since then I'd expect a peak at 938 MeV which isn't in your plots.
- something I ran into: the function log is base _e_, not base 10
- in that branch, both detectors 5555 and 101 are 'regular' detectors so they will skip low energy neutrals. however, even after enabling 'all' hits I don't get low energy hits.
```
--- a/geometry/hall/hallDaughter_merged.gdml
+++ b/geometry/hall/hallDaughter_merged.gdml
@@ -163,6 +163,7 @@
       <materialref ref="Vacuum"/>
       <solidref ref="cylHallDet_2"/>
       <auxiliary auxtype="SensDet" auxvalue="planeDet"/>
+      <auxiliary auxtype="DetType" auxvalue="all"/>
       <auxiliary auxtype="DetNo" auxvalue="101"/>
     </volume>
 
diff --git a/geometry/target/subTargetRegion.gdml b/geometry/target/subTargetRegion.gdml
index a77d2487..758e4ca8 100644
--- a/geometry/target/subTargetRegion.gdml
+++ b/geometry/target/subTargetRegion.gdml
@@ -241,6 +241,7 @@
     <solidref ref="skyshineVacuumDet1_solid"/>
     <auxiliary auxtype="Color" auxvalue="Blue"/>
     <auxiliary auxtype="SensDet" auxvalue="shyshineVacuumDet_1"/>
+    <auxiliary auxtype="DetType" auxvalue="all"/>
     <auxiliary auxtype="DetNo" auxvalue="5555"/>
   </volume>
```

In summary, can confirm this issue exists. Now I'd probably want to look at the prex sim too. Do you have a link?

**Comment by cipriangal** [comment]
*plot is hit.p 
*logX scale is base 10 made by this method: https://github.com/JeffersonLab/remoll/blob/feature-add-SBSbunker/analysis/hallRad/hallRad.cc#L354
*similar issues were seen in 5540 https://github.com/JeffersonLab/remoll/blob/feature-add-SBSbunker/geometry/target/subTargetRegion.gdml#L260 is DetType="all" needed or is secondaries and lowenergyneutrals enough?
*prexSim: https://github.com/cipriangal/prexSim .. doesn't do anything really. it just processes all hits afaik.

**Comment by wdconinc** [comment]
DetType="all" is identical to adding both secondaries *and* lowenergyneutrals (but it does not include optical photons). I figured that all of the above was true. Any difference when running kryptonite vs no kryptonite?

**Comment by cipriangal** [comment]
haven't tried that yet.

**Comment by wdconinc** [comment]
It's... a... bit... slow.

**Comment by wdconinc** [comment]
No meaningful impact of disabling kryptonite (other than slowing down). Still no hits < 10^-2 MeV.

**Comment by wdconinc** [comment]
Also with geant4.10.02-p02 there are no thermal neutrons.

**Comment by wdconinc** [comment]
Were the prexsim simulations with a hydrogen target or with a nuclear target? Could it be that with the beam generator we just don't expect that many thermal neutrons? I just can't find any difference. Regardless of geant4 version (4.10.02-p02 vs 4.10.05) or physics list (QGSP_BERT_HP vs QGSP_BERT). And I can't or think of other cutoffs (energy cutoffs, step limits). All hits should get stored in the detector.

**Comment by cipriangal** [comment]
it was with a nuclear target (Pb), but the remoll output on the roof (det 101) also doesn't show any thermals .. and at that point you have neutrons passing through lots of material resulting in thermalization. One thing we can try to do is run few MeV neutrons into the target instead of electrons and see if they thermalize either in the target or outside of the shielding.

**Comment by wdconinc** [comment]
Yeah, I tried turning the lh2 into full-on aluminum and dumped the 11 GeV electeon beam in it, but with the HP physics list it would crash reliably after about 40k events (another bug to track down).

**Comment by wdconinc** [comment]
Confirmed that all hits in det 101 are indeed saved to file, no hits are skipped.

**Comment by cipriangal** [comment]
could it be that we are somehow recording the wrong energy?! .. the 10keV peak there really looks like a thermal peak .. which should be 4 orders of magnitude down (~1eV).

**Comment by wdconinc** [comment]
Confirmed today that:
- low energy neutrons are simulated when using QGSP_BERT_HP,
- they are stored in the output ROOT files,
- but hit.p does not correspond to kinetic energy at these energies, so we have to be careful what we plot: we should plot `sqrt(hit.p*hit.p-hit.m*hit.m)-hit.m` instead. Once we plot that we do get hits down to the eV level.

Action items:
- Explicitly include hit.K as kinetic energy (hit.E remains total energy, hit.p remains momentum),
- Also found bug in remollGenericDetector for low-energy neutrals, which checks if total energy < 0.1 MeV. This turned that selection useless since even electrons have a total energy > 0.511 MeV. Likely will remove the lowenergyneutrals flag entirely.
